libraries/reconstruction_cuda/cuda_gpu_geo_transformer.cpp

```diff
@@ -454,45 +241,24 @@ void GeoTransformer<T>::loadCoefficients(const Matrix1D<T> &X,
 
 template<typename T>
 template<typename T_IN>
-void GeoTransformer<T>::produceAndLoadCoeffs(int splineDegree,
-        const MultidimArray<T_IN> &input) {
+void GeoTransformer<T>::produceAndLoadCoeffs(
+    const MultidimArray<T_IN> &input) {
+
+    // if no cast is needed, then don't copy input to impIn
+    const MultidimArray<T> * input_ptr;
     MultidimArray<T> tmpIn;
-    typeCast(input, tmpIn);
-    // copy data to gpu
+
+    if (std::is_same< T, T_IN >::value) {
```

Comment:
this is not necessary. See typeCast implementation, it does it for you.

Reply:
typeCast always copies data. This solution doesn't copy if T and T_IN are same. I measured it and just omitting this copy achieved higher speedup than writing the new kernel.

Reply:
You're right. I will ask if this is proper behaviour